### PR TITLE
Derivation Error - Baseline Mouthwash Collected

### PR DESCRIPTION
**Pull request by abhinavjonnada82**
Title^^

This pull request addresses the following issue: [issue #707](https://github.com/episphere/connect/issues/707).

Issue Recreated:
When a user accidentally checks mouthwash or urine collection. On the background, derived variable & timestamps get updated in participants table. When the user goes back to uncheck the collection & returns to review screen. The derived variable & timestamps don't get updated when unchecked.
[Link to Video for Issue Recreation](https://www.loom.com/share/10e903ebd2af413eaf8adea4e8c51dee?sid=aecd7b23-c16f-4982-9f23-ee3d3ddd2c05)

Proof of Concept (PoC):

When a user performs a check and uncheck on collections, it should trigger an update to derived variables and timestamps accordingly. 
[Link to Video for PoC](https://www.loom.com/share/9a7745643ec34822a855040bc139ecbe?sid=28e5d0cd-ec68-4ad7-967e-85f0c4c471a6)

Note:
Only urine and mouthwash derived variables require an update on check and uncheck on collection. Blood is skipped for the time being since there are multiple tubes, and the derived variable for blood appears to be correct.

### CODE REVIEW

**Review by bransteitterbr — APPROVED** [review]

File: src/shared.js [hunk 1]
```diff
@@ -1431,17 +1432,21 @@ export const updateCollectionSettingData = async (biospecimenData, tubes, partic
...
     if (!settings[visit][conceptIds.urineCollectionSetting]) {
         urineTubes.forEach(tube => {
-            if(biospecimenData[tube.concept][conceptIds.collection.tube.isCollected] === conceptIds.yes) {
-
+            const tubeIsCollected = biospecimenData[tube.concept][conceptIds.collection.tube.isCollected] === conceptIds.yes;
+            const collectionSetting = biospecimenData[conceptIds.collection.collectionSetting] ;
+            const isResearch = collectionSetting === conceptIds.research;
+            const isClinical = collectionSetting === conceptIds.clinical;
+    
+            if(tubeIsCollected) {
```
Comment (Gbarra9) [inline src/shared.js:1443]:
```suggestion
            if (tubeIsCollected) {
```

File: src/shared.js [hunk 2]
```diff
@@ -1431,17 +1432,21 @@ export const updateCollectionSettingData = async (biospecimenData, tubes, partic
...
                 settings[visit][conceptIds.urineCollectionSetting] = biospecimenData[conceptIds.collection.collectionSetting];
 
-                if(biospecimenData[conceptIds.collection.collectionSetting] === conceptIds.research) {
+                if(isResearch) {
```
Comment (Gbarra9) [inline src/shared.js:1446]:
```suggestion
                if (isResearch) {
```

File: src/shared.js [hunk 3]
```diff
@@ -1431,17 +1432,21 @@ export const updateCollectionSettingData = async (biospecimenData, tubes, partic
...
                     settings[visit][conceptIds.baseline.urineCollectedTime] = biospecimenData[conceptIds.collection.collectionTime];
                 }
-                else if(biospecimenData[conceptIds.collection.collectionSetting] === conceptIds.clinical) {
+                else if(isClinical) {
```
Comment (Gbarra9) [inline src/shared.js:1449]:
```suggestion
                else if (isClinical) {
```

File: src/shared.js [hunk 4]
```diff
@@ -1431,17 +1432,21 @@ export const updateCollectionSettingData = async (biospecimenData, tubes, partic
...
         });
     }
-        
+
     if (!settings[visit][conceptIds.urineCollectionSetting]) {
         urineTubes.forEach(tube => {
-            if(biospecimenData[tube.concept][conceptIds.collection.tube.isCollected] === conceptIds.yes) {
-
+            const tubeIsCollected = biospecimenData[tube.concept][conceptIds.collection.tube.isCollected] === conceptIds.yes;
+            const collectionSetting = biospecimenData[conceptIds.collection.collectionSetting] ;
```
Comment (Gbarra9) [inline src/shared.js:1439]:
```suggestion
            const collectionSetting = biospecimenData[conceptIds.collection.collectionSetting];
```

File: src/shared.js [hunk 5]
```diff
@@ -1454,25 +1459,71 @@ export const updateCollectionSettingData = async (biospecimenData, tubes, partic
...
 
-                if(biospecimenData[conceptIds.collection.collectionSetting] === conceptIds.research) {
+    if (!settings[visit][conceptIds.mouthwashCollectionSetting]) {
+        mouthwashTubes.forEach(tube => {
+            const isTubeCollected = biospecimenData[tube.concept][conceptIds.collection.tube.isCollected] === conceptIds.yes;
+            const collectionSetting = biospecimenData[conceptIds.collection.collectionSetting];
+            const isResearch = collectionSetting === conceptIds.research;
+            if(isTubeCollected) {
+                settings[visit][conceptIds.mouthwashCollectionSetting] = collectionSetting;
+                if(isResearch) {
```
Comment (Gbarra9) [inline src/shared.js:1497]:
```suggestion
            if (isTubeCollected) {
                settings[visit][conceptIds.mouthwashCollectionSetting] = collectionSetting;
                if (isResearch) {
```

File: src/shared.js [hunk 6]
```diff
@@ -1454,25 +1459,71 @@ export const updateCollectionSettingData = async (biospecimenData, tubes, partic
...
             }
         });
     }
+    else if (settings[visit][conceptIds.baseline.mouthwashCollectedTime] !== '' && participantData[conceptIds.baseline.mouthwashCollected] === conceptIds.yes){
+        mouthwashTubes.forEach(tube => {
+            const isTubeNotCollected = biospecimenData[tube.concept][conceptIds.collection.tube.isCollected] === conceptIds.no;
+            const isParticipantMouthwashCollected = participantData[conceptIds.baseline.mouthwashCollected] === conceptIds.yes;
+            if(isTubeNotCollected &&  isParticipantMouthwashCollected) {
+                delete settings[visit][conceptIds.mouthwashCollectionSetting]
+                if(biospecimenData[conceptIds.collection.collectionSetting] === conceptIds.research) {
```
Comment (Gbarra9) [inline src/shared.js:1509]:
```suggestion
            if (isTubeNotCollected &&  isParticipantMouthwashCollected) {
                delete settings[visit][conceptIds.mouthwashCollectionSetting]
                if (biospecimenData[conceptIds.collection.collectionSetting] === conceptIds.research) {
```

File: src/shared.js [hunk 7]
```diff
@@ -1454,25 +1459,71 @@ export const updateCollectionSettingData = async (biospecimenData, tubes, partic
...
+            const collectionSetting = biospecimenData[conceptIds.collection.collectionSetting];
+            const isResearch = collectionSetting === conceptIds.research;
+            const isClinical = collectionSetting === conceptIds.clinical;
+
+            if(tubeIsNotCollected && participantUrineCollected) {
+                delete settings[visit][conceptIds.urineCollectionSetting];
+                if(isResearch) {
+                    delete settings[visit][conceptIds.baseline.urineCollectedTime];
+                }
+                else if(isClinical) {
```
Comment (Gbarra9) [inline src/shared.js:1475]:
```suggestion
            if (tubeIsNotCollected && participantUrineCollected) {
                delete settings[visit][conceptIds.urineCollectionSetting];
                if (isResearch) {
                    delete settings[visit][conceptIds.baseline.urineCollectedTime];
                }
                else if (isClinical) {
```

File: src/shared.js [hunk 8]
```diff
@@ -1454,25 +1459,71 @@ export const updateCollectionSettingData = async (biospecimenData, tubes, partic
...
+                    settings[visit][conceptIds.clinicalDashboard.urineCollected] = conceptIds.no;
+                    delete settings[visit][conceptIds.clinicalDashboard.urineCollectedTime];
+                    settings[visit][conceptIds.anySpecimenCollected] = conceptIds.no;
 
-    if (!settings[visit][conceptIds.mouthwashCollectionSetting]) {
-        mouthwashTubes.forEach(tube => {
-            if(biospecimenData[tube.concept][conceptIds.collection.tube.isCollected] === conceptIds.yes) {
+                    if(!(settings[visit][conceptIds.anySpecimenCollectedTime])) {
```
Comment (Gbarra9) [inline src/shared.js:1480]:
```suggestion
                    if (!(settings[visit][conceptIds.anySpecimenCollectedTime])) {
```

**Review by Gbarra9 — APPROVED** [review]

File: src/shared.js [hunk 9]
```diff
@@ -1431,17 +1432,21 @@ export const updateCollectionSettingData = async (biospecimenData, tubes, partic
...
         urineTubes.forEach(tube => {
-            if(biospecimenData[tube.concept][conceptIds.collection.tube.isCollected] === conceptIds.yes) {
-
+            const tubeIsCollected = biospecimenData[tube.concept][conceptIds.collection.tube.isCollected] === conceptIds.yes;
+            const collectionSetting = biospecimenData[conceptIds.collection.collectionSetting] ;
+            const isResearch = collectionSetting === conceptIds.research;
+            const isClinical = collectionSetting === conceptIds.clinical;
+    
+            if(tubeIsCollected) {
                 settings[visit][conceptIds.urineCollectionSetting] = biospecimenData[conceptIds.collection.collectionSetting];
```
Comment (anthonypetersen) [inline src/shared.js:1444]: if you're going to declare `collectionSetting` you might as well use it here

File: src/shared.js [hunk 10]
```diff
@@ -1454,25 +1459,71 @@ export const updateCollectionSettingData = async (biospecimenData, tubes, partic
...
+
+            if(tubeIsNotCollected && participantUrineCollected) {
+                delete settings[visit][conceptIds.urineCollectionSetting];
+                if(isResearch) {
+                    delete settings[visit][conceptIds.baseline.urineCollectedTime];
+                }
+                else if(isClinical) {
+                    settings[visit][conceptIds.clinicalDashboard.urineCollected] = conceptIds.no;
+                    delete settings[visit][conceptIds.clinicalDashboard.urineCollectedTime];
+                    settings[visit][conceptIds.anySpecimenCollected] = conceptIds.no;
```
Comment (anthonypetersen) [inline src/shared.js:1478]: is this still valid if a blood tube has been collected?

File: src/shared.js [hunk 11]
```diff
@@ -1454,25 +1459,71 @@ export const updateCollectionSettingData = async (biospecimenData, tubes, partic
             }
         });
     }
+    else if (settings[visit][conceptIds.baseline.urineCollectedTime] !== '' ||  settings[visit][conceptIds.clinicalDashboard.urineCollectedTime] !== ''){
+        urineTubes.forEach(tube => {
+            const tubeIsNotCollected = biospecimenData[tube.concept][conceptIds.collection.tube.isCollected] === conceptIds.no;
+            const participantUrineCollected = participantData[conceptIds.baseline.urineCollected] === conceptIds.yes;
+            const collectionSetting = biospecimenData[conceptIds.collection.collectionSetting];
```
Comment (anthonypetersen) [inline src/shared.js:1466]: can collection setting be declared outside of these statements so it doesn't have to keep getting declared?

File: src/shared.js [hunk 12]
```diff
@@ -1454,25 +1459,71 @@ export const updateCollectionSettingData = async (biospecimenData, tubes, partic
...
+    }
 
-                settings[visit][conceptIds.mouthwashCollectionSetting] = biospecimenData[conceptIds.collection.collectionSetting];
 
-                if(biospecimenData[conceptIds.collection.collectionSetting] === conceptIds.research) {
+    if (!settings[visit][conceptIds.mouthwashCollectionSetting]) {
+        mouthwashTubes.forEach(tube => {
+            const isTubeCollected = biospecimenData[tube.concept][conceptIds.collection.tube.isCollected] === conceptIds.yes;
+            const collectionSetting = biospecimenData[conceptIds.collection.collectionSetting];
+            const isResearch = collectionSetting === conceptIds.research;
```
Comment (anthonypetersen) [inline src/shared.js:1494]: see comment about declaring variables in high scope

File: src/shared.js [hunk 13]
```diff
@@ -1454,25 +1459,71 @@ export const updateCollectionSettingData = async (biospecimenData, tubes, partic
...
+        })
+    }
 
     const settingData = {
         [conceptIds.collectionDetails]: settings,
         uid: participantData?.state?.uid
     };
+
+    // update derived variable if specimens unchecked after accidently checking them & derived varibales set to yes
+    if (Object.keys(derivedVariables).length !== 0) {
```
Comment (anthonypetersen) [inline src/shared.js:1523]: just add `...derivedVariables` to the end of the declaration for settingData

File: src/shared.js (hunk 10 above)
Comment (anthonypetersen) [reply src/shared.js:1477]: Let me rephrase... If a blood tube is collected and a urine tube is accidentally collected, when we go back and uncheck the urine tube wouldn't we still want `anySpecimenCollected` to be `yes` ?

**Review by anthonypetersen — APPROVED** [review]
